src/ActionsApple.cpp

```diff
@@ -0,0 +1,120 @@
+#include "Actions.hpp"
+#include "MemAllocHeap.hpp"
+#include "StatCache.hpp"
+#include "BinaryData.hpp"
+
+#if defined(TUNDRA_APPLE)
+
+#include <copyfile.h>
+#include <sys/time.h>
+#include <sys/stat.h>
+
+ExecResult CopyFiles(const FrozenFileAndHash* src_files, const FrozenFileAndHash* target_files, size_t files_count, StatCache* stat_cache, MemAllocHeap* heap)
+{
+    ExecResult result;
+    memset(&result, 0, sizeof(result));
+    char tmpBuffer[1024];
+
+    for(size_t i = 0; i < files_count; ++i)
+    {
+        const char* src_file = src_files[i].m_Filename;
+        const char* target_file = target_files[i].m_Filename;
+
+        FileInfo src_file_info = StatCacheStat(stat_cache, src_file);
+        if (!src_file_info.Exists())
+        {
+            result.m_ReturnCode = -1;
+            snprintf(tmpBuffer, sizeof(tmpBuffer), "The source path %s does not exist.", src_file);
+            break;
+        }
+
+        if (src_file_info.IsDirectory())
+        {
+            result.m_ReturnCode = -1;
+            snprintf(tmpBuffer, sizeof(tmpBuffer), "The source path %s is a directory, which is not supported.", src_file);
+            break;
+        }
+
+        FileInfo dst_file_info = StatCacheStat(stat_cache, target_file);
+        if (dst_file_info.Exists())
+        {
+            if (dst_file_info.IsDirectory())
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The target path %s already exists as a directory.", target_file);
+                break;
+            }
+
+            if (dst_file_info.IsReadOnly())
+            {
+                result.m_ReturnCode = -1;
+                snprintf(tmpBuffer, sizeof(tmpBuffer), "The target path %s already exists and is read-only.", target_file);
+                break;
+            }
+        }
+
+        copyfile_state_t state = copyfile_state_alloc();
+        result.m_ReturnCode = copyfile(src_file, target_file, state, COPYFILE_ALL | COPYFILE_UNLINK | COPYFILE_CLONE | COPYFILE_DATA_SPARSE);
```

Comment:
Will this preserve symlinks? Should we use COPYFILE_NOFOLLOW ?

Reply:
Yes, the `COPYFILE_CLONE` flag implies `COPYFILE_NOFOLLOW_SRC`.